python/paddle/fluid/tests/unittests/test_elementwise_gradient_op.py

```diff
@@ -0,0 +1,102 @@
+import unittest
+import numpy as np
+
+import paddle.fluid.core as core
+import paddle.fluid as fluid
+
+
+class TestElementWiseAddOp(unittest.TestCase):
+    def __assert_close(self, tensor, np_array, msg, atol=1e-4):
+        self.assertTrue(np.allclose(np.array(tensor), np_array, atol=atol), msg)
+
+    def check_forward_backward(self):
+        def test_with_place(place):
+            out_grad = np.random.random_sample(self.x.shape).astype(np.float32)
+            x_grad = out_grad
+            sum_axis = range(0, len(self.x.shape))
+            del sum_axis[self.axis]
+            y_grad = np.sum(out_grad, axis=tuple(sum_axis))
+
+            var_dict = locals()
+            var_dict['y'] = self.y
+            var_dict['x'] = self.x
+            var_dict['out'] = self.out
+            var_dict['y@GRAD'] = y_grad
+            var_dict['x@GRAD'] = x_grad
+            var_dict['out@GRAD'] = out_grad
+
+            var_names = ['x', 'y', 'out', 'y@GRAD', 'x@GRAD', 'out@GRAD']
+            ground_truth = {name: var_dict[name] for name in var_names}
+
+            program = fluid.Program()
+            with fluid.program_guard(program):
+                block = program.global_block()
+                for name in ground_truth:
+                    block.create_var(
+                        name=name,
+                        dtype='float32',
+                        shape=ground_truth[name].shape)
+                elementwise_add_op = block.append_op(
+                    type="elementwise_add",
+                    inputs={
+                        "X": block.var('x'),
+                        "Y": block.var('y'),
+                    },
+                    outputs={"Out": block.var('out'), })
+
+                # generate backward op_desc
+                grad_op_desc_list, op_grad_to_var = core.get_grad_op_desc(
+                    elementwise_add_op.desc, set(), [])
+                grad_op_desc = grad_op_desc_list[0]
+                new_op_desc = block.desc.append_op()
+                new_op_desc.copy_from(grad_op_desc)
+                for var_name in grad_op_desc.output_arg_names():
+                    block.desc.var(var_name.encode("ascii"))
+                grad_op_desc.infer_var_type(block.desc)
+                grad_op_desc.infer_shape(block.desc)
+                for arg in grad_op_desc.output_arg_names():
+                    grad_var = block.desc.find_var(arg.encode("ascii"))
+                    grad_var.set_dtype(core.VarDesc.VarType.FP32)
+
+                exe = fluid.Executor(place)
+                out = exe.run(program,
+                              feed={
+                                  name: var_dict[name]
+                                  for name in ['x', 'y', 'out@GRAD']
+                              },
+                              fetch_list=['x@GRAD', 'y@GRAD'])
+                self.__assert_close(x_grad, out[0], "x@GRAD")
+                self.__assert_close(y_grad, out[1], "y@GRAD", atol=2.0)
```

Comment:
atol=2.0 ? So large?

Reply:
The process of getting gradient involves accumulative operation and the shape of input is too large, x.shape = [2,32,220,220], y.shape = [32], so the diff of the result between Python and C++(CUDA) is bigger.

```
Python result
array([96771.625, 96801.26 , 96760.63 , 96893.32 , 96946.375, 96911.95 ,
       96837.16 , 96716.88 , 96931.305, 96746.445, 96781.81 , 96827.22 ,
       97108.19 , 96866.8  , 96894.17 , 96776.36 , 96720.61 , 96992.8  ,
       96641.664, 96772.305, 96698.16 , 96675.64 , 96805.86 , 96710.6  ,
       96733.37 , 96858.41 , 96771.516, 97033.516, 96820.48 , 96726.09 ,
       96784.66 , 96740.76 ], dtype=float32)

C++ result
array([96771.1  , 96801.05 , 96761.125, 96892.69 , 96946.73 , 96910.93 ,
       96837.516, 96716.27 , 96930.99 , 96748.18 , 96781.41 , 96828.016,
       97108.17 , 96866.62 , 96893.836, 96775.74 , 96720.98 , 96992.24 ,
       96642.36 , 96771.14 , 96698.195, 96675.055, 96805.88 , 96711.305,
       96732.086, 96858.97 , 96770.734, 97033.92 , 96819.95 , 96725.72 ,
       96784.35 , 96740.76 ], dtype=float32)

CUDA result
array([96738.2  , 96891.58 , 96752.25 , 96972.39 , 96706.016, 96755.234,
       96731.42 , 96850.01 , 97060.125, 96790.375, 96587.12 , 96833.305,
       96709.3  , 96703.7  , 96842.61 , 96727.95 , 96966.586, 96791.33 ,
       97077.234, 96715.03 , 96850.26 , 96898.47 , 96780.97 , 96839.516,
       96835.61 , 96600.41 , 96517.08 , 96787.58 , 96758.37 , 96555.484,
       96882.94 , 96578.2  ], dtype=float32)
```
So the max diff between `Python result` and `C++ result` is 1.28125, the max diff between `Python result` and `CUDA result` is 0.015625.